server/src/main/java/org/spine3/server/reflect/MethodMap.java

```diff
@@ -37,20 +38,64 @@
  */
-public class MethodMap {
+public class MethodMap<H extends HandlerMethod> {
 
-    private final ImmutableMap<Class<? extends Message>, Method> map;
+    private final ImmutableMap<Class<? extends Message>, H> map;
 
-    public MethodMap(Class<?> clazz, Predicate<Method> filter) {
-        this(Methods.scan(clazz, filter));
+    private MethodMap(Class<?> clazz, HandlerMethod.Factory<H> factory) {
+        final Map<Class<? extends Message>, Method> rawMethods = scan(clazz, factory.getPredicate());
+
+        final ImmutableMap.Builder<Class<? extends Message>, H> builder = ImmutableMap.builder();
+        for (Map.Entry<Class<? extends Message>, Method> entry : rawMethods.entrySet()) {
+            final H value = factory.create(entry.getValue());
+            builder.put(entry.getKey(), value);
+        }
+
+        this.map = builder.build();
     }
 
-    private MethodMap(Map<Class<? extends Message>, Method> map) {
-        this.map = ImmutableMap.<Class<? extends Message>, Method>builder()
-                .putAll(map)
-                .build();
+    /**
+     * Creates a new method map for the passed class using the passed factory.
+     *
+     * @param clazz the class to inspect
+     * @param factory the factory for handler methods
+     * @param <H> the type of the handler methods
+     * @return new method map
+     */
+    public static <H extends HandlerMethod> MethodMap<H> create(Class<?> clazz, HandlerMethod.Factory<H> factory) {
+        return new MethodMap<>(clazz, factory);
+    }
+
+    /**
+     * Returns a map of the {@link HandlerMethod} objects to the corresponding message class.
+     *
+     * @param declaringClass   the class that declares methods to scan
+     * @param filter the predicate that defines rules for subscriber scanning
+     * @return the map of message subscribers
+     * @throws DuplicateHandlerMethodException if there are more than one handler for the same message class are encountered
+     */
+    private static Map<Class<? extends Message>, Method> scan(Class<?> declaringClass, Predicate<Method> filter) {
+        final Map<Class<? extends Message>, Method> tempMap = Maps.newHashMap();
+        for (Method method : declaringClass.getDeclaredMethods()) {
+            if (filter.apply(method)) {
+                final Class<? extends Message> messageClass = HandlerMethod.getFirstParamType(method);
+                if (tempMap.containsKey(messageClass)) {
+                    final Method alreadyPresent = tempMap.get(messageClass);
+                    throw new DuplicateHandlerMethodException(
+                            declaringClass,
+                            messageClass,
+                            alreadyPresent.getName(),
+                            method.getName());
+                }
+                tempMap.put(messageClass, method);
+            }
+        }
+        final ImmutableMap.Builder<Class<? extends Message>, Method> builder = ImmutableMap.builder();
```

Comment:
Could you refactor so that we don't use the temp map and add to the builder from the very beginning?